internal/storage/serde_events.go

```diff
@@ -0,0 +1,575 @@
+	"github.com/milvus-io/milvus/pkg/common"
+	"github.com/milvus-io/milvus/pkg/util/metautil"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+var _ RecordReader = (*compositeBinlogRecordReader)(nil)
+
+type compositeBinlogRecordReader struct {
+	blobs [][]*Blob
+
+	blobPos int
+	rrs     []array.RecordReader
+	closers []func()
+	fields  []FieldID
+
+	r compositeRecord
+}
+
+func (crr *compositeBinlogRecordReader) iterateNextBatch() error {
+	if crr.closers != nil {
+		for _, close := range crr.closers {
+			if close != nil {
+				close()
+			}
+		}
+	}
+	crr.blobPos++
+	if crr.blobPos >= len(crr.blobs[0]) {
+		return io.EOF
+	}
+
+	for i, b := range crr.blobs {
+		reader, err := NewBinlogReader(b[crr.blobPos].Value)
+		if err != nil {
+			return err
+		}
+
+		crr.fields[i] = reader.FieldID
+		// TODO: assert schema being the same in every blobs
+		crr.r.schema[reader.FieldID] = reader.PayloadDataType
+		er, err := reader.NextEventReader()
+		if err != nil {
+			return err
+		}
+		rr, err := er.GetArrowRecordReader()
+		if err != nil {
+			return err
+		}
+		crr.rrs[i] = rr
+		crr.closers[i] = func() {
+			rr.Release()
+			er.Close()
+			reader.Close()
+		}
+	}
+	return nil
+}
+
+func (crr *compositeBinlogRecordReader) Next() error {
+	if crr.rrs == nil {
+		if crr.blobs == nil || len(crr.blobs) == 0 {
+			return io.EOF
+		}
+		crr.rrs = make([]array.RecordReader, len(crr.blobs))
+		crr.closers = make([]func(), len(crr.blobs))
+		crr.blobPos = -1
+		crr.fields = make([]FieldID, len(crr.rrs))
+		crr.r = compositeRecord{
+			recs:   make(map[FieldID]arrow.Record, len(crr.rrs)),
+			schema: make(map[FieldID]schemapb.DataType, len(crr.rrs)),
+		}
+		if err := crr.iterateNextBatch(); err != nil {
+			return err
+		}
+	}
+
+	composeRecord := func() bool {
+		for i, rr := range crr.rrs {
+			if ok := rr.Next(); !ok {
+				return false
+			}
+			// compose record
+			crr.r.recs[crr.fields[i]] = rr.Record()
+		}
+		return true
+	}
+
+	// Try compose records
+	if ok := composeRecord(); !ok {
+		// If failed the first time, try iterate next batch (blob), the error may be io.EOF
+		if err := crr.iterateNextBatch(); err != nil {
+			return err
+		}
+		// If iterate next batch success, try compose again
+		if ok := composeRecord(); !ok {
+			// If the next blob is empty, return io.EOF (it's rare).
+			return io.EOF
+		}
+	}
+	return nil
+}
+
+func (crr *compositeBinlogRecordReader) Record() Record {
+	return &crr.r
+}
+
+func (crr *compositeBinlogRecordReader) Close() {
+	for _, close := range crr.closers {
+		if close != nil {
+			close()
+		}
+	}
+}
+
+func parseBlobKey(blobKey string) (colId FieldID, logId UniqueID) {
+	if blobKey == "" {
+		return 0, 0
+	}
+	if _, _, _, colId, logId, ok := metautil.ParseInsertLogPath(blobKey); ok {
+		return colId, logId
+	}
+	if colId, err := strconv.ParseInt(blobKey, 10, 64); err == nil {
+		// data_codec.go generate single field id as blob key.
+		return colId, 0
+	}
+	return -1, -1
```

Comment:
nit: there is a const InvalidUniqueID, `return InvalidUniqueID, InvalidUniqueID`